File: services/frontend/crm-ui/src/utils/postalCodeRules.ts
export interface PostalCodeRule {
  country: string
  regex: RegExp
  hint: string
}

const FALLBACK_RULE: PostalCodeRule = {
  country: 'Other',
  regex: /^[A-Za-z0-9][A-Za-z0-9 -]{3,9}$/,
  hint: '4-10 letters, numbers, spaces, or hyphens',
}

const RULES: PostalCodeRule[] = [
  { country: 'Singapore', regex: /^\d{6}$/, hint: '6 digits (e.g. 238801)' },
  { country: 'United States', regex: /^\d{5}(?:-\d{4})?$/, hint: '12345 or 12345-6789' },
  {
    country: 'United Kingdom',
    regex: /^[A-Z]{1,2}\d[A-Z\d]?\s?\d[A-Z]{2}$/i,
    hint: 'UK postcode format (e.g. SW1A 1AA)',
  },
  { country: 'Canada', regex: /^[A-Za-z]\d[A-Za-z][ -]?\d[A-Za-z]\d$/, hint: 'A1A 1A1' },
  { country: 'Australia', regex: /^\d{4}$/, hint: '4 digits (e.g. 3000)' },
  { country: 'Germany', regex: /^\d{5}$/, hint: '5 digits (e.g. 10115)' },
  { country: 'France', regex: /^\d{5}$/, hint: '5 digits (e.g. 75001)' },
  { country: 'India', regex: /^\d{6}$/, hint: '6 digits (e.g. 110001)' },
  { country: 'Japan', regex: /^\d{3}-?\d{4}$/, hint: '123-4567 or 1234567' },
  FALLBACK_RULE,
]

export const COUNTRY_OPTIONS = RULES.map(rule => rule.country)

export function getPostalCodeRule(country: string): PostalCodeRule {
  const normalized = country.trim().toLowerCase()
  return RULES.find(rule => rule.country.toLowerCase() === normalized) ?? FALLBACK_RULE
}

export function isPostalCodeValidForCountry(country: string, postalCode: string): boolean {
  const normalizedPostalCode = postalCode.trim()
  if (
    !normalizedPostalCode ||
    normalizedPostalCode.length < 4 ||
    normalizedPostalCode.length > 10
  ) {
    return false
  }
  return getPostalCodeRule(country).regex.test(normalizedPostalCode)
}

File: services/backend/client/src/main/java/com/scroogebank/crm/client_service/validation/PostalCodeRules.java
package com.scroogebank.crm.client_service.validation;

import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Country-specific postal code validation rules used by client create/update flows.
 */
public final class PostalCodeRules {
	private static final Rule FALLBACK_RULE = new Rule(
		"Other",
		Pattern.compile("^[A-Za-z0-9][A-Za-z0-9 -]{3,9}$")
	);

	private static final List<Rule> RULES = List.of(
		new Rule("Singapore", Pattern.compile("^\\d{6}$")),
		new Rule("United States", Pattern.compile("^\\d{5}(?:-\\d{4})?$")),
		new Rule("United Kingdom", Pattern.compile("^[A-Z]{1,2}\\d[A-Z\\d]?\\s?\\d[A-Z]{2}$", Pattern.CASE_INSENSITIVE)),
		new Rule("Canada", Pattern.compile("^[A-Za-z]\\d[A-Za-z][ -]?\\d[A-Za-z]\\d$")),
		new Rule("Australia", Pattern.compile("^\\d{4}$")),
		new Rule("Germany", Pattern.compile("^\\d{5}$")),
		new Rule("France", Pattern.compile("^\\d{5}$")),
		new Rule("India", Pattern.compile("^\\d{6}$")),
		new Rule("Japan", Pattern.compile("^\\d{3}-?\\d{4}$")),
		FALLBACK_RULE
	);

	private PostalCodeRules() {}

	public static String validate(String country, String postalCode) {
		String normalizedCountry = country == null ? "" : country.trim();
		String normalizedPostalCode = postalCode == null ? "" : postalCode.trim();

		if (normalizedCountry.isEmpty()) {
			return "Country is required";
		}
		if (normalizedPostalCode.isEmpty()) {
			return "Postal code is required";
		}
		if (normalizedPostalCode.length() < 4 || normalizedPostalCode.length() > 10) {
			return "Postal code must be 4-10 characters";
		}

		Rule rule = ruleForCountry(normalizedCountry);
		if (!rule.pattern().matcher(normalizedPostalCode).matches()) {
			return "Postal code must match " + rule.country() + " format";
		}
		return null;
	}

	private static Rule ruleForCountry(String country) {
		String normalized = country.trim().toLowerCase(Locale.ROOT);
		return RULES.stream()
			.filter(rule -> rule.country().toLowerCase(Locale.ROOT).equals(normalized))
			.findFirst()
			.orElse(FALLBACK_RULE);
	}

	private record Rule(String country, Pattern pattern) {}
}

